### PR TITLE
CLN refactor core dtypes

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Some refactorings found by Sourcery https://sourcery.ai/

I've removed the ones of the kind
```diff
- if param:
-     var = a
- else:
-     var = b
+ var = a if param else b
```

### CODE REVIEW

**Comment by jreback** [comment]
look ok. these may have broken things.

**Comment by MarcoGorelli** [comment]
Indeed - the refactoring
```diff
-    if len(types) == 0:
+    if not types:
```
was wrong, as `types` could be a `Series` here (as in the test `pandas/tests/arrays/sparse/test_accessor.py::test_to_coo`), in which was `not types` raises
```
ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

Have fixed up and changed the annotation

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -1535,7 +1539,7 @@ def maybe_cast_to_datetime(value, dtype: DtypeObj, errors: str = "raise"):
     return value
 
 
-def find_common_type(types: List[DtypeObj]) -> DtypeObj:
+def find_common_type(types: Union[List[DtypeObj], Series]) -> DtypeObj:
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:1542]: where are we passing a Series? we shouldn't do this

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/dtypes/cast.py:1542]: in `pandas/tests/arrays/sparse/test_accessor.py::test_to_coo`

Comment (MarcoGorelli) [reply pandas/core/dtypes/cast.py:1542]: I'll dig deeper at the weekend

Comment (MarcoGorelli) [reply pandas/core/dtypes/cast.py]: @jreback the problem is that when we do
```
        df = pd.DataFrame({"A": [0, 1, 0], "B": [1, 0, 0]}, dtype="Sparse[int64, 0]")
        result = df.sparse.to_coo()
```
then, in `to_coo` from pandas/core/arrays/sparse/accessor.py, we have
```
        dtype = find_common_type(self._parent.dtypes)
```
and so this is how we pass a Series (`self._parent.dtypes` is a Series).

Is there a better way to fix this than
```
        dtype = find_common_type(self._parent.dtypes.to_list())
```
?

**Comment by jreback** [comment]
can you merge master ping on green

**Comment by MarcoGorelli** [comment]
> can you merge master ping on green

Green, barring (probably unrelated?) travis error:
```
____________________________ test_deprecated_kwargs ____________________________

[gw0] linux -- Python 3.7.8 /home/travis/miniconda3/envs/pandas-dev/bin/python

    def test_deprecated_kwargs():

        df = pd.DataFrame({"A": [2, 4, 6], "B": [3, 6, 9]}, index=[0, 1, 2])

        buf = df.to_json(orient="split")

        with tm.assert_produces_warning(FutureWarning):

>           tm.assert_frame_equal(df, read_json(buf, "split"))

pandas/tests/io/json/test_deprecated_kwargs.py:15: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <contextlib._GeneratorContextManager object at 0x7fc4d8ce5110>

type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):

        if type is None:

            try:

>               next(self.gen)

E               AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed <ssl.SSLSocket fd=12, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.20.0.66', 55192), raddr=('52.216.144.227', 443)>"), '/home/travis/build/pandas-dev/pandas/pandas/core/dtypes/base.py', 435), ('ResourceWarning', ResourceWarning("unclosed file <_io.BufferedWriter name='/tmp/tmpnm7c5epi.xls'>"), '/home/travis/build/pandas-dev/pandas/pandas/core/dtypes/base.py', 435), ('ResourceWarning', ResourceWarning("unclosed file <_io.BufferedWriter name='/tmp/tmplgig23ur.xlsx'>"), '/home/travis/build/pandas-dev/pandas/pandas/core/dtypes/base.py', 435)]

../../../miniconda3/envs/pandas-dev/lib/python3.7/contextlib.py:119: AssertionError
```

File: pandas/core/arrays/sparse/accessor.py [hunk 2]
```diff
@@ -329,7 +329,7 @@ def to_coo(self):
         import_optional_dependency("scipy")
         from scipy.sparse import coo_matrix
 
-        dtype = find_common_type(self._parent.dtypes)
+        dtype = find_common_type(self._parent.dtypes.to_list())
```
Comment (MarcoGorelli) [inline pandas/core/arrays/sparse/accessor.py:332]: this is because `find_commont_type` from `pandas/core/dtypes/cast.py` requires a list

**Comment by jreback** [comment]
moving off 1.2 until ready

**Comment by MarcoGorelli** [comment]
Sure, no problem (though as far as I can tell, this is ready - anything else need changing?)

**Comment by jreback** [comment]
> Sure, no problem (though as far as I can tell, this is ready - anything else need changing?)

I think it was fine, just want to make sure after merge

**Comment by MarcoGorelli** [comment]
sure, have merged

File: pandas/core/dtypes/missing.py [hunk 3]
```diff
@@ -484,11 +485,11 @@ def _array_equivalent_object(left, right, strict_nan):
...
                     return False
             except TypeError as err:
-                if "Cannot compare tz-naive" in str(err):
+                if "Cannot compare tz-naive" in str(
+                    err
+                ) or "boolean value of NA is ambiguous" in str(err):
                     # tzawareness compat failure, see GH#28507
                     return False
-                elif "boolean value of NA is ambiguous" in str(err):
-                    return False
```
Comment (jbrockmendel) [inline pandas/core/dtypes/missing.py:491]: i think this is for coverage

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MarcoGorelli

**Pull request merged**